### PR TITLE
fix(docbuild): Fix Doxygen warnings caused by format errors

### PR DESCRIPTION
**Pull request by vwheeler63**
### Description of the feature or fix

Fix Doxygen warnings where Doxygen does not recognize some parts of the documentation string, causing it to be omitted in most cases in Doxygen output.

Fixes Doxygen formatting part of issue lvgl#6480.

### Notes
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.  (n/a)
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.  (n/a)
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.  (n/a)
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).  (n/a)
- Run `scripts/code-format.py` ([astyle](http://astyle.sourceforge.net/install.html) version [v3.4.12](https://github.com/szepeviktor/astyle/releases/tag/v3.4.12) needs to be installed) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html).  (n/a)
- Mark the Pull request as [Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request) while you are working on the first version, and mark is as _Ready_ when it's ready for review.  Done.
- When changes were requested, [re-request review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to notify the maintainers.  Done.
- Help us to review this Pull Request! Anyone can [approve or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).  (n/a)

### CODE REVIEW

**Review by kisvegabor — DISMISSED** [review]
Looks good, thank you Vic!

**Comment by vwheeler63** [comment]
> Looks good, thank you Vic!

Message received!  **thumbs up**  I see you merged PR #6582, thank you.  But I do not see this PR merged with `master` yet.  If you will, give me a "ping" when this is done and I will update my fork and my local git tree.

I am hoping we can encourage Kevin to submit his PR (I see his changes -- I have one repo cloned from his LVGL repo, and the other (the one I am working on) cloned from the master LVGL repo, so I can compare them side by side.  And I do not yet see his PR (with lots of changes to .PY and .RST files, and from what I can see, 201 Doxygen formatting changes in the LVGL ./src/ directory that he made to fix errors that were showing up between `breathe` and `sphinx`.  (These do not conflict with my changes, but rather supplement the additional 134 changes I made.)

@kdschlosser Kevin, what say you?  :-)

Kind regards,
Vic

**Comment by kdschlosser** [comment]
#6542 is the PR. 

There are 201 files changed across 38 commits.

**Comment by vwheeler63** [comment]
@kdschlosser @kisvegabor 

> #6542 is the PR.
> 
> There are 201 files changed across 38 commits.

My apologies!  I see you're still working on it. (Good job keeping it up to date with `lvgl/master` branch to make Gabor's job easier.)

Meanwhile, I have figured out how to get those changes into a local branch so that I can work with them.  And for Gabor, I can confirm there were no merge conflicts, it's generating a nice clean build now except for the (mostly) expected Doxygen warnings I get when I turn on the default warnings.

**Review by XuNeo — DISMISSED** [review]

**Review by liamHowatt — DISMISSED** [review]

**Comment by kdschlosser** [comment]
as soon as this branch has it's merge conflicts sorted out and merged I am going to sync the PR I have and see how the output looks. I also have to mess about with it not catching errors correctly.

**Comment by vwheeler63** [comment]
Hi @kisvegabor !  I compared this branch with current `lvgl/lvgl` master branch to resolve the 2 conflicts.  It should be "good to go" now.

**Comment by vwheeler63** [comment]
> as soon as this branch has it's merge conflicts sorted out and merged I am going to sync the PR I have and see how the output looks. I also have to mess about with it not catching errors correctly.

@kdschlosser I hear you.  Added time, but I agree it was a good idea for me not to try to merge into your branch.  You had a lot more going on there than I initially realized.  If you want, I can take these changes and cherry pick and send you a PR if that helps for your local work?

**Comment by kdschlosser** [comment]
It's not a big deal for me to wait until this PR gets merged. The majority of the changes I made are in the ReST files so fixing any merge conflicts shouldn't be that big of a deal because most if not all of the changes made in this PR reside in the C header files. So there shouldn't be too much to deal with.

**Comment by vwheeler63** [comment]
> It's not a big deal for me to wait until this PR gets merged.

@kdschlosser Okee dokee.  Just wanted to make sure we were on the same page.  :-)

File: src/drivers/x11/lv_x11.h
Comment (XuNeo) [inline src/drivers/x11/lv_x11.h:46]: What is the correct way to mark function in generated doc? So we can click it to jump to the function doc.

**Review by XuNeo — APPROVED** [review]

**Review by kisvegabor — APPROVED** [review]

**Comment by kisvegabor** [comment]
Hi Vic,

I've just noticed that there some more merge conflicts. It's manly because of the change from `_lv_*` to `lv_*`.

@XuNeo I'll be on a business trip this week so might be slower to respond.
@liamHowatt Please take a look at this issue once the merge conflict is fixed, and approve it if it looks good.

File: src/drivers/x11/lv_x11.h
Comment (kdschlosser) [reply src/drivers/x11/lv_x11.h:46]: This is not something that is not easily supported from the doxygen code. I have started looking at another python library that handles the C docstrings since Breathe is not being maintained any longer. The one I am looking at doesn't use Doxygen at all.. It also supports sphinx ReST syntax for the docstrings in the C header files. It is actively being maintained which is a good thing but more importantly is it supports both doxygen style docstrings and also ReST ones so we can transition if it was something we were going to use.

I am going to drop the author of the library to see if there would be interest in expanding the library so it will parse doxygen style docstrings and be able to also parse sphinx directives that have been added to those docstrings.

**Comment by kisvegabor** [comment]
> I found a possible replacement for breathe. Is it OK if I open a dialog with the author of the library so we can go over what we would like to see done and if the library would be a good fit or not?

If it's maintained, works and not very new I have no objection.

**Comment by kdschlosser** [comment]
it's been around for 7 years. so not that new. 

It appears to be maintained, There are not many issues opened for problems, that can either be a good thing or it could mean it isn't something that gets used a lot. don't know yet. 

I will open a dialog and see where it goes.

File: src/drivers/x11/lv_x11.h
Comment (vwheeler63) [reply src/drivers/x11/lv_x11.h:46]: @XuNeo @kdschlosser 

> What is the correct way to mark function in generated doc? So we can click it to jump to the function doc.

Good question!  Actually, Doxygen handles this automatically (adding the link automatically) by leaving the function name _bare_ with parens at the end.  It does the same with bare data types (provides a link to their documentation).  From what I have seen, once Doxygen creates the link, then `breathe` and `sphinx` preserve it I believe.  ( @kdschlosser correct me if I am wrong on that point.)  Just make sure it is spelled the same as the actual function in source code, like this:

```
 * - create keyboard (lv_x11_keyboard_create())
 * - create mouse (with scrollwheel, lv_x11_mouse_create() and lv_x11_mousewheel_create())
```

Here is a current working example from elsewhere (in a Doxygen comment):

```
 * - At run time, initialize each list and list item before use using
 *   + OCTWLL_InitList(), and
 *   + and OCTWLL_InitItem() (respectively).
```

produces this in the Doxygen output (this is a screen shot so the links are not live):

image

P.S.  Don't bother going back and fixing these as I have have fixed about a dozen of these examples in the batch I am currently working on.  I will catch all of them -- no problem.

--Vic

**Comment by vwheeler63** [comment]
@kisvegabor 

Hi, Gabor!  

Regarding

> I've just noticed that there some more merge conflicts. It's manly because of the change from _lv_* to lv_*.

when I loaded this notification this morning, I found this, so perhaps another merge fixed the conflict?

image

Kind regards,
Vic

File: src/drivers/x11/lv_x11.h
Comment (kdschlosser) [reply src/drivers/x11/lv_x11.h:46]: @vwheeler63 

The answer is, it depends., It depends on what you are wanting to link to. If it is something that resides in the doxygen domain then it will be preserved. you cannot link to a specific place in the documentation that resides in one of the ReST files. You also want be able to link to a specific header file, this is because the documentation representation of the header file resides in the sphinx domain.

In sphinx (ReST) I am able to make a connection to the doxygen domain by using directives along the lines of 
```rest
:c:macro:`SOME_MACRO_NAME`
```

and I am able to reference specific header files by using

```rest
:ref:`header_name_h`
```

and I am able to link to sections of the documentation like the quick overview by using

```rest
:ref:`quick-overview`
```

There is no way to link to the quick overview from a doxygen docstring. YET... I am working on being able to use ReST syntax in the doxygen docstrings. Having this ability would help out a lot.

Comment (vwheeler63) [reply src/drivers/x11/lv_x11.h:46]: >  The answer is, it depends.

Nice.

Doxygen will link to a file by just naming it _bare_ in a Doxygen comment.

Doxygen will also allow you to link to separate places within the Doxygen "realm", like if you create a link (several ways to do that, including the `@anchor  my_anchor_name` command) -- you can reference anywhere inside the same Doxygen "realm" by

```
@ref  my_anchor_name
```

and it will go there (in any file that was part of that Doxygen run).

One thing it _does_ require, however, is that there is documentation for the "thing" being linked to.  I forgot how many different kinds of links are automatic.  For sure there is filenames, global variable names, typedefs, function names, and macro names.

But for sure (agreed) it's not going to know anything about any anchors in .RST files.

File: src/misc/lv_style.h
Comment (liamHowatt) [inline src/misc/lv_style.h:32]: Are these actually causing a warning? I've seen the `/**` convention used in many places in C code.

**Review by liamHowatt — APPROVED** [review]
Indeed it seems there are no conflicts. Maybe kisvegabor just needed to refresh?

File: src/misc/lv_style.h
Comment (kdschlosser) [reply src/misc/lv_style.h:32]: The warnings were suppressed in the documentation build system. @vwheeler63 has enabled them so now all of the warnings get shown by doxygen. Some things that produce warnings and have docstrings will still get documented but other will produce warnings but the documentation doesn't populate into the final output. Ideally there should be zero warnings and zero errors even if the build completes. That is the goal. I don't have a great amount of experience with doxygen and I didn't notice that the warnings had been suppressed. This was done prior to me making any changes to the documentation build system. Someone else had done that prior to me doing any work on LVGL.

@vwheeler63 has far more experience with doxygen that I do and he found that the warnings had been suppressed. and when they were turned back on there was a HUGE number that ended up being produced.

Comment (liamHowatt) [reply src/misc/lv_style.h:32]: Got it, so the changes are guided by vwheeler63's Doxygen warnings. I see the discussion in the issue now.

Comment (vwheeler63) [reply src/misc/lv_style.h:32]: @liamHowatt @XuNeo 

Hi, Liam!

Additionally, turning Doxygen warnings ON (just the default ones so far) shows us where there was _meant to be_ documentation that was NOT making it into the eventual API documentation generated by the documentation build system, and it was hard to detect because the MEMBERS were there (e.g. macros, functions, structs, struct members), but the intended documentation strings (or "docstrings" in Python lingo) were silently not showing up!

And in what you are asking about:  "/\*\*" => "/\*" -- in that case, a string that was meant to label a GROUP of items was being erroneously interpreted (by Doxygen) as a "docstring" for the FIRST item in a list, which of course did not make sense when it showed up only labeling the first item in the generated API documentation.  This was not a Doxygen error, but an error in formatting of the comments.  So after a discussion about this with Gabor, we settled on having Doxygen NOT bring in that string as a "docstring" since it was mis-labeling 1 item only rather than a group.  There are ways to group things in Doxygen but we reserved that for a later project.  The change from 2 "\*\*" to 1 "\*" forces Doxygen to EXCLUDE that string from the input -- intentionally and with coordination with Gabor.  There are around 6-12 of these (by memory) in this PR and the PRs about to be submitted.

The 2 asterisks are signals to Doxygen to use that string to generate documentation for something -- normally the "thing" that is _right after_ it.

"/\*\*< comment_text \*/" tells Doxygen to use that string as documentation for the "thing" just _before_ it.

Comment (liamHowatt) [reply src/misc/lv_style.h:32]: I see, thanks for the explanation! Maybe that convention came from Doxygen but people forgot that it came from there and has significance when used in a Doxygen context.

Comment (vwheeler63) [reply src/misc/lv_style.h:32]: > I see, thanks for the explanation! Maybe that convention came from Doxygen but people forgot that it came from there and has significance when used in a Doxygen context.

Agreed!  Part of this plan is to do a pretty good write-up in the Coding Styles part of the documentation so that people can learn what is needed and wanted.

Comment (kdschlosser) [reply src/misc/lv_style.h:32]: Yessir that is the game plan.

I have been having a good conversation with the author of Hawkmoth which is something I am looking at as a replacement for breathe. Breathe is no longer being developed unless the author is paid to do it. So it needs to be replaced with something that is still being developed. The thing I like about Hawkmoth is it will parse doxygen docstrings without using doxygen. It uses clang and attaches to the compiler using it's C API. This is done using ctypes in Python. The other thing I really like about it is not only is it able to read doxygen docstrings but it will also read sphinx formatted ReST docstrings as well. Nice thing about this is we will have the ability to create connections between the API documentation and things located in the sphinx documentation which is something that is not able to currently be done. It also allows for documented sections in the C header files. So things like the widget documentation that is currently located in the docs folder is able to be relocated into the C header files for each widget. I can organize things however it is wanting to be done but the big thing is the documentation is able to be located with the code instead of being a separate file that needs to be updated.

Comment (kdschlosser) [reply src/misc/lv_style.h:32]: My point to that last comment is in order for it to work properly the doxygen syntax needs to be correct otherwise it is going to have issues.

Comment (vwheeler63) [reply src/misc/lv_style.h:32]: > the big thing is the documentation is able to be located with the code

I think this is important as well -- without it, it makes it that much harder to learn about how to use LVGL.  By the way, Doxygen supports this, but I don't know the data path, so not sure if feasible/workable or not (yet).

Comment (vwheeler63) [reply src/misc/lv_style.h:32]: @liamHowatt 

Hi, Liam!

In this context, is it you or me who is responsible to press the [Resolve Conversation] button?

--Vic

**Comment by kisvegabor** [comment]
Oh, yeah, maybe it just needed a refresh. 
Merged now! :tada:

**Comment by vwheeler63** [comment]
> Oh, yeah, maybe it just needed a refresh. Merged now! 🎉

@kisvegabor 

Yeay!  Thanks for looking in, even though you're busy.  :-)

File: src/misc/lv_style.h
Comment (liamHowatt) [reply src/misc/lv_style.h:32]: There is no hard rule. I personally only resolve conversations if I started the conversation, the issue is resolved, and there isn't an ongoing discussion. If someone decides to resolve it sooner that's fine.

Comment (vwheeler63) [reply src/misc/lv_style.h:32]: > There is no hard rule. I personally only resolve conversations if I started the conversation, the issue is resolved, and there isn't an ongoing discussion. If someone decides to resolve it sooner that's fine.

Cool.  Thanks!  :-)

**Pull request merged**
